capsules/src/mac.rs

```diff
@@ -0,0 +1,365 @@
+//! Implements 802.15.4 MAC device functionality as an abstraction layer over
+//! the raw radio transceiver hardware. The abstraction difference between a MAC
+//! device and a raw radio transceiver is that the MAC devices exposes a
+//! frame-oriented interface to its users, whereas the radio transceiver
+//! transmits raw byte sequences. There is some abstraction breaking here,
+//! though because the following are still implemented at the hardware level:
+//! - CSMA-CA backoff
+//! - FCS generation and verification
+//!
+//! TODO: Encryption/decryption
+//! TODO: Sending beacon frames
+//! TODO: Channel scanning
+
+use core::cell::Cell;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::radio;
+use net::ieee802154::*;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct FrameInfo {
+    // These offsets are relative to buf[radio::PSDU_OFFSET..] so that
+    // the mac frame length is data_offset + data_len
+    mac_payload_offset: usize,
+    data_offset: usize,
+    data_len: usize,
+    security_params: Option<(Security, [u8; 16])>,
+}
+
+impl FrameInfo {
+    pub fn frame_length(&self) -> usize {
+        self.data_offset + self.data_len
+    }
+
+    pub fn remaining_data_capacity(&self, buf: &[u8]) -> usize {
+        buf.len() - radio::PSDU_OFFSET - radio::MFR_SIZE - self.data_offset - self.data_len
+    }
+
+    pub fn append_payload(&mut self, buf: &mut [u8], payload: &[u8]) -> ReturnCode {
+        if payload.len() > self.remaining_data_capacity(buf.as_ref()) {
+            return ReturnCode::ENOMEM;
+        }
+        let begin = radio::PSDU_OFFSET + self.data_offset + self.data_len;
+        buf[begin..begin + payload.len()].copy_from_slice(payload);
+        self.data_len += payload.len();
+
+        ReturnCode::SUCCESS
+    }
+}
+
+pub trait Mac {
```

Comment:
Document. But also...

Why does this need to be a trait? Is the idea that you may have several different MAC implementations for 802.15.4? If so, it seems better for this to be in `radio.rs` (for now), since it'd be odd to implement a different MAC that depends on this MAC just for the trait.

It's probably also OK to get rid of this for now.

One possible utility for it is if you wanted to mock the MAC layer in order to test higher layers.